### PR TITLE
feat: Validation for JSON models

### PR DESCRIPTION
**Pull request by d-jeffery**
## Description

- resolves https://github.com/openfga/language/issues/100

## Review Checklist
- [ ] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [ ] The correct base branch is being used, if not `main`
- [ ] I have added tests to validate that the change in functionality is working as expected

### CODE REVIEW

File: tests/data/json-validation-cases.yaml [hunk 1]
```diff
@@ -0,0 +1,3020 @@
+---
+- name: valid schema version
+  json: |
+    {
+      "schema_version": "1.1"
```
Comment (rhamzeh) [inline tests/data/json-validation-cases.yaml:5]: use 1.3, because 1.2 will soon become valid :)

File: pkg/js/tests/validator.test.ts [hunk 2]
```diff
@@ -72,3 +77,41 @@ describe("validateDSL", () => {
     });
   });
 });
+
+describe("validateJSON", () => {
+  const testCases = loadInvalidJSONSValidationTestCases();
```
Comment (rhamzeh) [inline pkg/js/tests/validator.test.ts:82]: can we append the DSL ones after converting to these being run?

File: tests/data/json-validation-cases.yaml (hunk 1 above)
Comment (rhamzeh) [reply tests/data/json-validation-cases.yaml:5]: There is no need to have everything in here, as many can be written in DSL format and imported.

That has the additional benefit of not needing to write many of the tests twice.

Only json that is not supported by the DSL needs to have extra tests

**Review by rhamzeh — COMMENTED** [review]
Any chance we can have `validateDSL` call `validateJSON` but with the dslString optional?

From a quick skim the main difference between the two is that `validateDSL` passes the index.

So validateJSON can do that if the DSL string metadata is provided to the function, otherwise it passes undefined. WDYT?

**Comment by d-jeffery** [comment]
> Any chance we can have `validateDSL` call `validateJSON` but with the dslString optional?
> 
> From a quick skim the main difference between the two is that `validateDSL` passes the index.
> 
> So validateJSON can do that if the DSL string metadata is provided to the function, otherwise it passes undefined. WDYT?

Thats what I thought of first;

Problem I see is that every function under validateJSON expects the `lines: string[]` to be passed to it to generate a line number and col for exceptions. This is consumed by the error collector, and before every exception a `lineIndex` is calculated using `lines: string[]`.

It would mean making all the functions that currently expect `lines` and/or `lineIndex` to make the parameter `string[] | undefined` and/or `number | undefined`, then handle the undefined behavior.

e.g Of a typical exception generation

```typescript
function populateRelations(
  lines: string[],
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  collector: any,
  authorizationModel: AuthorizationModel,
  typeRegex: ValidationRegex,
  relationRegex: ValidationRegex,
) {
  // Looking at the types
  authorizationModel.type_definitions?.forEach((typeDef) => {
    const typeName = typeDef.type;

    if (typeName === Keyword.SELF || typeName === ReservedKeywords.THIS) {
      const lineIndex = getTypeLineNumber(typeName, lines);
      collector.raiseReservedTypeName(typeName, lineIndex);
    }

    if (!typeRegex.regex.test(typeName)) {
      const lineIndex = getTypeLineNumber(typeName, lines);
      collector.raiseInvalidName(typeName, typeRegex.rule, undefined, lineIndex);
    }
```    

This is main reason i split the code - if you'd prefer handling undefined parameters here, I can try that. I have already made line & col values optional for the collector, so that behavior is already working.

File: pkg/js/util/exceptions.ts [hunk 3]
```diff
@@ -228,12 +240,27 @@ const createDuplicateRelationError = (props: BaseProps, relationName: string) =>
 
 const createDuplicateRelationshipDefinitionError = (props: BaseProps) => {
   const { errors, lines, lineIndex, symbol } = props;
+
+  console.dir(props);
```
Comment (rhamzeh) [inline pkg/js/util/exceptions.ts:244]: 🔪

File: pkg/js/validator/validate-dsl.ts [hunk 4]
```diff
@@ -684,54 +695,59 @@ function modelValidation(
...
+      if (!relationRegex.regex.test(relationName)) {
         const typeIndex = getTypeLineNumber(typeName, lines);
         const lineIndex = getRelationLineNumber(relationName, lines, typeIndex);
-        collector.raiseInvalidName(lineIndex, relationName, relationRegex.rule, typeName);
-      } else if (encounteredRelationsInType[relationName]) {
+        collector.raiseInvalidName(relationName, relationRegex.rule, typeName, lineIndex);
+      }
+
+      // TODO: Will never be true, consider removing
+      if (encounteredRelationsInType[relationName]) {
```
Comment (rhamzeh) [inline pkg/js/validator/validate-dsl.ts:744]: 🔪

File: pkg/js/validator/validate-dsl.ts [hunk 5]
```diff
@@ -745,11 +761,11 @@ function populateRelations(
  * @param {ValidationOptions} options
  */
 export function validateJSON(
-  jsonString: string,
   authorizationModel: AuthorizationModel,
   options: ValidationOptions = {},
+  jsonString?: string,
```
Comment (rhamzeh) [inline pkg/js/validator/validate-dsl.ts:766]: dslString

**Review by rhamzeh — APPROVED** [review]
All good from my end! You think we can rename the file from `validate-dsl` to `validate`?

**Review by ewanharris — APPROVED** [review]

**Pull request merged**